src/Handler/MailCryptKeyHandler.php

```diff
@@ -0,0 +1,173 @@
+     */
+    public function toPkcs8(string $privateKey): string
+    {
+        // Unfortunately, there doesn't seem to be a way to transform elliptic curve
+        // keys from from traditional PEM to PKCS#8 format within PHP yet. The OpenSSL
+        // extension doesn't support PKCS#8 format and phpseclib doesn't support elliptic
+        // curves.
+        //
+        // Invoke `openssl pkey` system process to transform the EC key to PKCS#8 format
+
+        $process = new Process(['openssl', 'pkey', '-in', '-']);
+        $inputStream = new InputStream();
+        $process->setInput($inputStream);
+        $process->start();
+        $inputStream->write($privateKey);
+        $inputStream->close();
+        $process->wait();
+
+        sodium_memzero($privateKey);
+
+        return $process->getOutput();
+    }
+
+    /**
+     * @param User $user
+     *
+     * @throws \Exception
+     */
+    public function create(User $user): void
+    {
+        $pKey = openssl_pkey_new([
+            'private_key_type' => self::MAIL_CRYPT_PRIVATE_KEY_TYPE,
+            'curve_name' => self::MAIL_CRYPT_CURVE_NAME,
+        ]);
+        openssl_pkey_export($pKey, $privateKey);
+        $privateKey = base64_encode($this->toPkcs8($privateKey));
+        $keyPair = new MailCryptKeyPair($privateKey, base64_encode(openssl_pkey_get_details($pKey)['key']));
+        sodium_memzero($privateKey);
+
+        // get plain user password
+        if (null === $plainPassword = $user->getPlainPassword()) {
+            throw new \Exception('plainPassword should not be null');
+        }
+
+        $mailCryptPrivateSecret = CryptoSecretHandler::create($keyPair->getPrivateKey(), $plainPassword);
+        $user->setMailCryptPublicKey($keyPair->getPublicKey());
+        $user->setMailCryptPrivateSecret($mailCryptPrivateSecret->encode());
+        $user->setPlainMailCryptPrivateKey($keyPair->getPrivateKey());
+
+        // Clear variables with confidential content from memory
+        $keyPair->erase();
+        sodium_memzero($plainPassword);
+
+        $this->manager->flush();
+    }
+
+    /**
+     * @param User   $user
+     * @param string $oldPlainPassword
+     *
+     * @throws \Exception
+     */
+    public function update(User $user, string $oldPlainPassword): void
+    {
+        // get plain user password to be encrypted
+        if (null === $plainPassword = $user->getPlainPassword()) {
+            throw new \Exception('plainPassword should not be null');
+        }
+
+        if (null === $secret = $user->getMailCryptPrivateSecret()) {
+            throw new \Exception('secret should not be null');
+        }
+
+        if (null === $privateKey = CryptoSecretHandler::decrypt(CryptoSecret::decode($secret), $oldPlainPassword)) {
+            throw new \Exception('decryption of mailCryptPrivateSecret failed');
+        }
+
+        $user->setMailCryptPrivateSecret(CryptoSecretHandler::create($privateKey, $plainPassword)->encode());
```

Comment:
I think you should re-use `updateWithPrivateKey` here?!

Reply:
Good catch, fixed!